src/i2c.rs

```diff
@@ -0,0 +1,369 @@
+
+impl Pins<I2C2>
+for (
+    PB10<Alternate<OpenDrain>>,
+    PB11<Alternate<OpenDrain>>,
+) {
+    const REMAP: bool = false;
+}
+
+/// I2C peripheral operating in master mode
+pub struct I2c<I2C, PINS> {
+    i2c: I2C,
+    pins: PINS,
+    bus_timeout: u32,
+    mode: Mode,
+    pclk1: u32,
+}
+
+impl<PINS> I2c<I2C1, PINS> {
+    pub fn i2c1(
+        i2c: I2C1,
+        pins: PINS,
+        mapr: &mut MAPR,
+        mode: Mode,
+        clocks: Clocks,
+        apb: &mut APB1,
+    ) -> Self
+        where
+            PINS: Pins<I2C1>,
+    {
+        mapr.mapr().modify(|_, w| w.i2c1_remap().bit(PINS::REMAP));
+        I2c::_i2c1(i2c, pins, mode, clocks, apb)
+    }
+}
+
+impl<PINS> I2c<I2C2, PINS> {
+    pub fn i2c2(
+        i2c: I2C2,
+        pins: PINS,
+        mode: Mode,
+        clocks: Clocks,
+        apb: &mut APB1,
+    ) -> Self
+        where
+            PINS: Pins<I2C2>,
+    {
+        I2c::_i2c2(i2c, pins, mode, clocks, apb)
+    }
+}
+
+
+macro_rules! busy_wait {
+    ($cycles:expr, $i2c:expr, $flag:ident) => {
+        {
+            let mut cycles_left = $cycles;
+            loop {
+                if cycles_left == 0 {
+                    break Err(Error::Timeout)
+                } else {
+                    cycles_left -= 1;
+                }
+
+                let sr1 = $i2c.sr1.read();
+
+                if sr1.berr().bit_is_set() {
+                    break Err(Error::Bus);
+                } else if sr1.arlo().bit_is_set() {
+                    break Err(Error::Arbitration);
+                } else if sr1.af().bit_is_set() {
+                    break Err(Error::Acknowledge);
+                } else if sr1.ovr().bit_is_set() {
+                    break Err(Error::Overrun);
+                } else if sr1.$flag().bit_is_set() {
+                    break Ok(());
+                } else {
+                    // try again
+                }
+            }
+        }
+    }
+}
+
+macro_rules! hal {
+    ($($I2CX:ident: ($i2cX:ident, $i2cXen:ident, $i2cXrst:ident),)+) => {
+        $(
+            impl<PINS> I2c<$I2CX, PINS> {
+                /// Configures the I2C peripheral to work in master mode
+                pub fn $i2cX(
+                    i2c: $I2CX,
+                    pins: PINS,
+                    mode: Mode,
+                    clocks: Clocks,
+                    apb: &mut APB1,
+                ) -> Self {
+                    apb.enr().modify(|_, w| w.$i2cXen().enabled());
+                    apb.rstr().modify(|_, w| w.$i2cXrst().set_bit());
+                    apb.rstr().modify(|_, w| w.$i2cXrst().clear_bit());
+
+                    // 25 ms bus timeout @ ~10 clks per busy_wait cycle
+                    let bus_timeout = (clocks.sysclk().0 / 400).max(10);
+                    let pclk1 = clocks.pclk1().0;
+
+                    assert!(mode.get_frequency() <= 400_000);
+
+                    let mut i2c = I2c { i2c, pins, bus_timeout, mode, pclk1 };
+                    i2c.init();
```

Comment:
I'd rather compute proper values for the registers and save them here instead of doing every time we `init()` after communication errors, but I'll lose the typesafe register API then.